### PR TITLE
Advanced Search Page API Performance TLC

### PR DESCRIPTION
**Pull request by tony-sappe**
[DEV-909](https://federal-spending-transparency.atlassian.net/browse/DEV-909)

### High level description
A number of changes over the past few months took a toll on Advanced Search Page API performance. Noteworthy changes include:
- Using matviews for downloads of data pre-FY2008
- Including loan subsidy costs in the aggregation amounts for visualization API endpoints

This PR primarily makes changes to materialized view indexes and ORM code to craft slightly different SQL select statements.

### Technical details
1. All matviews which are not necessary for downloads were changed back to FY2008 and forward only (`summary_*` views)
2. The matviews which do include FY2001-FY2007 data had substantial rework to their indexes (`universal_*` views)
    - All indexes are partial indexes using action date (existing indexes are now only for >= '2007-10-01')
    - New indexes which cover older data for downloads
3. `universal_award_matview` is now ordered by `total_obligation DESC NULLS LAST` instead of `action_date DESC`
4. No longer using fiscal_year for queries
5. Using `generated_pragmatic_obligation` instead of federal_action_obligation or total_subsidy_cost depending on the award type
6. Capped Advanced Search Page API date range minimum start date to 2007-10-01
7. Combined action_date ranges when requested ranges overlap or are adjacent (e.g. {2008-10-01, 2009-09-30} and {2009-10-01, 2010-09-30} are merged into {2008-10-01, 2010-09-30} 
8. Explicit ORM->SQL ordering for spending by award to be DESC NULLS LAST which matches existing indexes

### The following are ALL required for the PR to be merged
- [x] Backend review completed
- [x] Matview impact assessment completed
- [x] Frontend impact assessment completed **N/A**
- [x] Data validation completed
- [x] API Performance evaluation completed and present

#### Matview impact assessment
All materialized views should be recreated as all have significant changes to either the view, or indexes, or both. Creation time should be shorter and total storage needs are reduced. Unfortunately neither are back to the original (FY2008+ only) metrics. Functionality won't break if the code is deployed before DB changes, but performance will degrade even more than currently.

#### Data validation completed
Aggregated obligation values need to be verified just to confirm accuracy, even thought the logic is the same, simply moved from API SQL to matview create SQL.

### CODE REVIEW

File: usaspending_api/core/validator/helpers.py [hunk 1]
```diff
@@ -16,9 +16,9 @@
 
 TINY_SHIELD_SEPARATOR = '|'
 
-INVALID_TYPE_MSG = 'Invalid value in \'{key}\'. \'{value}\' is not a valid {type}'
-ABOVE_MAXIMUM_MSG = 'Field \'{key}\' value \'{value}\' is above max {max}'
-BELOW_MINIMUM_MSG = 'Field \'{key}\' value \'{value}\' is below min {min}'
+INVALID_TYPE_MSG = 'Invalid value in \'{key}\'. \'{value}\' is not a valid ({type})'
```
Comment (dpb-bah) [inline usaspending_api/core/validator/helpers.py]: `[Value] is not a valid (string)`. Is that supposed to be `[Value] is not a valid type (string)`?

File: usaspending_api/database_scripts/matview_generator/universal_award_matview.json [hunk 2]
```diff
@@ -131,7 +131,7 @@
     "  (awards.category IS NOT NULL or contract_data.pulled_from='IDV') AND",
     "  latest_transaction.action_date >= '2000-10-01'",
     "ORDER BY",
-    "  latest_transaction.action_date DESC"
+    "  awards.total_obligation DESC NULLS LAST"
```
Comment (dpb-bah) [inline usaspending_api/database_scripts/matview_generator/universal_award_matview.json]: Asking more for documentation: why the change of order?

File: usaspending_api/search/tests/test_mock_data_search.py [hunk 3]
```diff
@@ -31,7 +31,7 @@ def all_filters():
         "place_of_performance_scope": "domestic",
         "place_of_performance_locations": [{"country": "USA"},
                                            {"country": "PQR"}],
-        "award_type_codes": ["A", "B", "02", '05', 'S'],
+        "award_type_codes": ["A", "B"],
```
Comment (dpb-bah) [inline usaspending_api/search/tests/test_mock_data_search.py]: Why the dropping of the "02", '05', 'S'?

File: usaspending_api/search/v2/views/search.py [hunk 4]
```diff
@@ -81,20 +84,28 @@ def post(self, request):
...
+                .values('award_type') \
+                .annotate(
+                    month=ExtractMonth('action_date'),
+                    year=ExtractYear('action_date'),
+                    transaction_amount=Sum('amount')) \
+                .values('month', 'year', 'transaction_amount')
         else:
-            data_set = queryset.values('fiscal_year')
-            if not (group == 'fy' or group == 'fiscal_year'):
+            queryset = queryset.values('fiscal_year')
```
Comment (dpb-bah) [inline usaspending_api/search/v2/views/search.py]: Can we include the `for Awards we use sum_transaction_amount` down here?

File: usaspending_api/settings.py [hunk 5]
```diff
@@ -19,6 +19,10 @@
 # User-specified timeout limit for streaming downloads
 DOWNLOAD_TIMEOUT_MIN_LIMIT = 10
 
+API_MAX_DATE = '2020-09-30'
```
[the remark this answers is not in the document]
Comment (tony-sappe) [reply usaspending_api/settings.py]: Yeah, I thought it was far enough into the future for now. It could be 2099-09-30 and it wouldn't change anything. It is for the date range combine logic

File: usaspending_api/search/v2/views/search.py (hunk 4 above)
Comment (tony-sappe) [reply usaspending_api/search/v2/views/search.py]: What is `sum_transaction_amount`?

File: usaspending_api/core/validator/helpers.py (hunk 1 above)
Comment (dpb-bah) [reply usaspending_api/core/validator/helpers.py]: It's just a grammatical error:
`' \'{value}\' is not a valid ({type})'`
will come out like `'A' is not a valid (int)'`
`\'{value}\' is not a valid type ({type})'` will come out like `'A' is not a valid type (int)'`

File: usaspending_api/search/v2/views/search.py (hunk 4 above)
Comment (dpb-bah) [reply usaspending_api/search/v2/views/search.py]: Sorry, this should just be `#  for Awards we use transaction_amount`.

Comment (dpb-bah) [reply usaspending_api/search/v2/views/search.py]: Transaction amount you can see below on lines 100 and 107.
